### PR TITLE
[vtgate] Add flag to pool connection read buffers

### PR DESCRIPTION
**Pull request by brirams**
## Description

We(Slack) found that when a vtgate is under heavy connection load, tail latencies of connect times
tend to spike into the tens of milliseconds. A contributing factor is GC slowdown that is
exacerbated by the repeated allocation of enw 16k read buffers for each incoming connection. This PR
adds a flag to the vtgate mysql server to enable pooling those read buffers and re-using them across
requests.

@henryr found that under a synthetic load, we were able to get connection times down by 55% at the
p999 and 40% at the p99:

Screen Shot 2022-08-16 at 10 37 16 AM

We've also run this against our loadtest environment and found, aside from less GC pauses, that our 
upstream clients saw a 20% reduction in connection failures, with a 5-8% drop in connect latencies 
across the board(p50 -> p999), and no perceived degradation in query latency.

This type of change has been [attempted in the past](https://github.com/vitessio/vitess/pull/4186)
and was closed out but since there has since been work to buffer connection writes, figured it made
sense to revisit that.

## Related Issue(s)

https://github.com/vitessio/vitess/pull/4186

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [ ] Tests were added or are not required
-   [ ] Documentation was added or is not required

## Deployment Notes

n/a

### CODE REVIEW

File: go/mysql/conn.go [hunk 1]
```diff
@@ -289,6 +301,13 @@ func (c *Conn) endWriterBuffering() error {
 	return c.bufferedWriter.Flush()
 }
 
+func (c *Conn) returnReader() {
+	if c.bufferedReader != nil {
+		c.bufferedReader.Reset(nil)
+		readersPool.Put(c.bufferedReader)
+	}
+}
+
```
Comment (harshit-gangal) [inline go/mysql/conn.go:310]: nit: can use guarded cause, simpler to read and less nested code.
```go
if c.bufferedReader == nil {
  return
}
c.bufferedReader.Reset(nil)
readersPool.Put(c.bufferedReader)
```

**Review by harshit-gangal — COMMENTED** [review]
can we run the benchmarks from here `go/mysql/query_benchmark_test.go` if it gives some additional insights?

**Comment by brirams** [comment]
Benchmark output from this branch(still need to compare with main but figured this was good to have for now):
```
 ~/n/d/vitess/go/mysql $ go test -v -bench=.
<elided>
goos: darwin
goarch: amd64
pkg: vitess.io/vitess/go/mysql
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkParallelShortQueries
BenchmarkParallelShortQueries-16     	       1	1428472710 ns/op	   0.00 MB/s
BenchmarkParallelMediumQueries
BenchmarkParallelMediumQueries-16    	       1	1482719290 ns/op	   0.01 MB/s
BenchmarkParallelRandomQueries
BenchmarkParallelRandomQueries-16    	       1	1467065363 ns/op
PASS
```

**Comment by brirams** [comment]
results off of the main branch:
```
~/n/d/vitess/go/mysql $ go test -v -bench=.
<elided>
goos: darwin
goarch: amd64
pkg: vitess.io/vitess/go/mysql
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkParallelShortQueries
BenchmarkParallelShortQueries-16     	       1	1588925544 ns/op	   0.00 MB/s
BenchmarkParallelMediumQueries
BenchmarkParallelMediumQueries-16    	       1	1576328375 ns/op	   0.01 MB/s
BenchmarkParallelRandomQueries
BenchmarkParallelRandomQueries-16    	       1	1627565065 ns/op
PASS
```

**Comment by brirams** [comment]
modified the benchmark to run with the read buffer pooling enable and disabled so we can compare more effectively:

```
~/n/d/vitess/go/mysql $ go test -v -bench=.
<elided>
goos: darwin
goarch: amd64
pkg: vitess.io/vitess/go/mysql
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkParallelShortQueries
BenchmarkParallelShortQueries-16                          	       1	1820589019 ns/op	   0.00 MB/s
BenchmarkParallelMediumQueries
BenchmarkParallelMediumQueries-16                         	       1	1798463863 ns/op	   0.01 MB/s
BenchmarkParallelRandomQueries
BenchmarkParallelRandomQueries-16                         	       1	1712101024 ns/op
BenchmarkParallelShortQueriesWithReadBufferPooling
BenchmarkParallelShortQueriesWithReadBufferPooling-16     	       1	1687282698 ns/op	   0.00 MB/s
BenchmarkParallelMediumQueriesWithReadBufferPooling
BenchmarkParallelMediumQueriesWithReadBufferPooling-16    	       1	1740339458 ns/op	   0.01 MB/s
BenchmarkParallelRandomQueriesWithReadBufferPooling
BenchmarkParallelRandomQueriesWithReadBufferPooling-16    	       1	1702365597 ns/op
PASS
```

File: go/vt/vtgate/plugin_mysql_server.go [hunk 2]
```diff
@@ -75,6 +75,8 @@ var (
 	mysqlConnWriteTimeout = flag.Duration("mysql_server_write_timeout", 0, "connection write timeout")
 	mysqlQueryTimeout     = flag.Duration("mysql_server_query_timeout", 0, "mysql query timeout")
 
+	mysqlConnBufferPooling = flag.Bool("mysql_server_pool_conn_read_buffers", false, "If set, the server will pool incoming connection read buffers")
```
Comment (harshit-gangal) [inline go/vt/vtgate/plugin_mysql_server.go:78]: new flags are to be added with `-` over `_` in the name

**Review by harshit-gangal — COMMENTED** [review]
changes looks good, just one flag naming convention change.

**Comment by harshit-gangal** [comment]
Would need to add in `release notes` for the new flag and what it means over https://github.com/vitessio/vitess/blob/main/doc/releasenotes/15_0_0_summary.md

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**